### PR TITLE
[RFC] grouping tensors in C++ for fused adam(w)

### PR DESCRIPTION
**Pull request by crcrpar**
This duplicates the core functionality of https://github.com/pytorch/pytorch/blob/master/torch/utils/_foreach_utils.py#L21 in C++.

The changes are
- manually setting `device_guard` and calling `device_check` in each fused kernel
- incrementing step counts and optionally decrementing them

Ideally there should be an equivalent in C++ which we can use for foreach optimizer and each foreach function so that we can remove the for-loop in Python. Note that once having each foreach function takes care of this grouping, the foreach optimizers could redundantly do the grouping; so still iterating over the lists of grouped tensors could make sense in foreach optimizers.

related: https://github.com/pytorch/pytorch/issues/58833, https://github.com/pytorch/pytorch/pull/89591#discussion_r1060267916

cc @ptrblck @ngimel

### CODE REVIEW

**Comment by albanD** [comment]
Nice!
Any number on perf improvement?

**Comment by crcrpar** [comment]
with huggingface/transformers example of gpt2-medium ([script](https://github.com/crcrpar/transformers/blob/868ec94fbe4b5b167e03fb958a32f6519dd8d932/examples/pytorch/language-modeling/run_clm.py))

master (2.0.0a0+git1767026)
```
***** train metrics *****
  epoch                    =        3.0
  train_loss               =     2.8572
  train_runtime            = 0:01:49.16
  train_samples            =       1200
  train_samples_per_second =     32.978
  train_steps_per_second   =      1.044
```

this pr
```
***** train metrics *****
  epoch                    =        3.0
  train_loss               =     2.8572
  train_runtime            = 0:01:39.96
  train_samples            =       1200
  train_samples_per_second =     36.014
  train_steps_per_second   =       1.14
```

machine: DGXStation (A100x4, AMD EPYC7742 x 64)

Command:
```
python run_clm.py --model_name_or_path gpt2-medium --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --per_device_train_batch_size 8 --per_device_eval_batch_size 8 --do_train --do_eval --output_dir /workspace/results --max_train_samples 1200 --warmup_steps 8 --overwrite_output_dir --optim adamw_torch
```

---

In the first run, this check https://github.com/pytorch/pytorch/blob/f9cc12eebdb0c6c9d8ab80ed2eebcd72e0a22244/torch/optim/adamw.py#L60
errored out so I removed it and things worked. So I'll remove.

**Comment by crcrpar** [comment]
t5-large

master
```
***** train metrics *****
  epoch                    =        1.0
  train_loss               =     2.6017
  train_runtime            = 0:00:39.23
  train_samples            =       5000
  train_samples_per_second =     127.43
  train_steps_per_second   =      1.019
```

this pr
```
***** train metrics *****
  epoch                    =        1.0
  train_loss               =     2.6017
  train_runtime            = 0:00:39.14
  train_samples            =       5000
  train_samples_per_second =    127.737
  train_steps_per_second   =      1.022
```

command:
```
python run_translation.py --model_name_or_path t5-large --do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 --max_source_length 512 --max_target_length 512 --num_train_epochs 1 --source_lang en --target_lang ro --dataset_name wmt16 --dataset_config ro-en --source_prefix "translate English to Romanian:"  --warmup_steps 50 --max_train_samples 5000 --dataloader_num_workers 2 --output_dir /workspace/result_t5-large --overwrite_output_dir --optim adamw_torch
```

**Comment by janeyx99** [comment]
Thanks for looking into this! Is your long term goal here to replace the python grouping util with a generalized C++ util or is this PR just for fusedAdamW?

**Comment by crcrpar** [comment]
> Is your long term goal here to replace the python grouping util with a generalized C++ util or is this PR just for fusedAdamW?

The former. that said, I'm not sure how to avoid multiple groupings for `_multi_tensor` optimizer functions at the moment

**Comment by janeyx99** [comment]
Yea, I agree with the concern if we were to expand foreach grouping to C++ as a generic util. 

I have a more general concern about adding grouping for general foreach ops (like add, mul, etc)--will this add significant perf cost for foreach calls where the tensors were _already_ properly grouped? What is the perf difference if we added grouping for every foreach op vs the existing checks we do today? Do you have some benchmarks comparing the existing checks logic vs the grouping logic?

If the difference is significant, would a potential solution be to allow users to pass in a skip_grouping flag that says "I'm sure I don't need grouping, just run + if it fails it's my fault"? This would require an API change though...which....is likely BC breaking. 

cc'ing @ngimel as well

**Comment by crcrpar** [comment]
> If the difference is significant, would a potential solution be to allow users to pass in a skip_grouping flag that says "I'm sure I don't need grouping, just run + if it fails it's my fault"? This would require an API change though...which....is likely BC breaking.

I had a similar thought: adding a flag to each foreach function and/or expose from cpp to python some function which takes list of `Tensor`s and returns `Dict[torch.device, Dict[torch.dtype, List[List[Tensor]]` which could be utilized by https://github.com/pytorch/pytorch/blob/c16b2916f15d7160c0254580f18007eb0c373abc/torch/cuda/amp/grad_scaler.py#L195 as well

**Comment by janeyx99** [comment]
I would agree that having a separate C++ util exposed that could be optionally called from relevant endpoints would be better than changing the API haha. Do you have the numbers for how long the grouping would take with the gpt2-medium params?

**Comment by crcrpar** [comment]
some more numbers (top: this pr, bottom: master)

"Self" numbers look good. 

### FP32 - the same gpt2 script above

```
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                              Optimizer.step#AdamW.step         0.14%      21.845ms         0.46%      72.387ms       3.619ms       0.000us         0.00%     161.130ms       8.056ms            20  
                                    aten::_fused_adamw_         0.04%       6.821ms         0.32%      50.472ms       1.262ms     161.130ms         1.68%     161.130ms       4.028ms            40  
                    Optimizer.zero_grad#AdamW.zero_grad         0.09%      14.349ms         5.52%     865.262ms      43.263ms       0.000us         0.00%      19.468ms     973.400us            2

---

                              Optimizer.step#AdamW.step         0.19%      29.690ms         0.46%      71.838ms       3.592ms       0.000us         0.00%     161.619ms       8.081ms            20  
                                    aten::_foreach_add_         0.02%       2.650ms         0.02%       3.294ms      82.350us     180.000us         0.00%     180.000us       4.500us            40  
                                    aten::_fused_adamw_         0.02%       3.216ms         0.25%      38.790ms     969.750us     161.439ms         1.69%     161.439ms       4.036ms            40  
                    Optimizer.zero_grad#AdamW.zero_grad         0.09%      14.145ms         5.53%     867.559ms      43.378ms       0.000us         0.00%      19.454ms     972.700us            2
```


### FP16 AMP (autocast & GradScaler)

```
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                              Optimizer.step#AdamW.step         0.06%       8.618ms         0.13%      19.301ms     965.050us       0.000us         0.00%     162.830ms       8.142ms            20
                                    aten::_fused_adamw_         0.06%       8.422ms         0.07%      10.665ms     266.625us     162.830ms         1.91%     162.830ms       4.071ms            40
                    Optimizer.zero_grad#AdamW.zero_grad         0.10%      14.414ms         0.36%      52.794ms       2.640ms       0.000us         0.00%      19.235ms     961.750us            20

---

                              Optimizer.step#AdamW.step         0.13%      18.071ms         0.20%      28.279ms       1.414ms       0.000us         0.00%     162.600ms       8.130ms            20
                                    aten::_foreach_add_         0.02%       2.582ms         0.02%       3.014ms      75.350us     160.000us         0.00%     160.000us       4.000us            40  
                                    aten::_fused_adamw_         0.02%       2.775ms         0.03%       4.242ms     106.050us     162.197ms         1.90%     162.197ms       4.055ms            40  
                                    aten::_foreach_sub_         0.02%       2.478ms         0.02%       2.938ms      73.450us     243.000us         0.00%     243.000us       6.075us            40
                    Optimizer.zero_grad#AdamW.zero_grad         0.10%      13.828ms         0.37%      53.763ms       2.688ms       0.000us         0.00%      19.232ms     961.600us            20
```

File: aten/src/ATen/native/cuda/fused_adam_utils.cuh [hunk 1]
```diff
@@ -166,6 +173,85 @@ struct FusedAdamMathFunctor {
...
+      ref.device() == t.device() &&
+      t.layout() == at::kStrided &&
+      t.is_non_overlapping_and_dense() &&
+      ref.sizes() == t.sizes() &&
+      ref.strides() == t.strides();
+  };
+  for (const auto & tensor_index : c10::irange(num_tensors)) {
+    const auto & first_tensor = nested_tensorlists[0][tensor_index];
+    const auto device = nested_tensorlists[0][tensor_index].device();
+    const auto scalar_type = nested_tensorlists[0][tensor_index].scalar_type();
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/fused_adam_utils.cuh:224]: nit or i could just be wrong, but why not
```suggestion
    const auto device = first_tensor.device();
    const auto scalar_type = first_tensor.scalar_type();
```

File: aten/src/ATen/native/cuda/fused_adam_utils.cuh [hunk 2]
```diff
@@ -166,6 +173,85 @@ struct FusedAdamMathFunctor {
...
+    for (const auto & i : c10::irange(num_lists)) {
+      const auto only_device_check = has_state_steps && i == num_lists - 1;
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/fused_adam_utils.cuh:226]: is `only_device_check` simply `is_step_tensor`? nit on a more indicative name if so

File: aten/src/ATen/native/cuda/fused_adam_utils.cuh [hunk 3]
```diff
@@ -166,6 +173,85 @@ struct FusedAdamMathFunctor {
...
+      const auto & t = nested_tensorlists[i][tensor_index];
+      if (only_device_check) {
+        TORCH_CHECK(t.is_cuda() && device == t.device());
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/fused_adam_utils.cuh:229]: We'd probably need a special mechanism for excepting step tensors in the generic function

File: aten/src/ATen/native/cuda/fused_adam_utils.cuh [hunk 4]
```diff
@@ -166,6 +173,85 @@ struct FusedAdamMathFunctor {
...
+      } else {
+        TORCH_CHECK(is_tensor_okay(nested_tensorlists[0][tensor_index], nested_tensorlists[i][tensor_index]));
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/fused_adam_utils.cuh:231]:
```suggestion
        TORCH_CHECK(is_tensor_okay(first_tensor, nested_tensorlists[i][tensor_index]));
```

File: aten/src/ATen/native/cuda/fused_adam_utils.cuh (hunk 3 above)
Comment (janeyx99) [reply aten/src/ATen/native/cuda/fused_adam_utils.cuh:229]: Ah, is it fair to say that in the generic grouping, we will just skip this for loop check? In other words, this check is only applicable for optimizer steps?

**Comment by janeyx99** [comment]
The numbers do look good! I have some questions:
1. Is it correct that this PR moved the grouping logic onto CUDA by calling it within the kernel? If so, are the numbers for fused_adam(w) basically showing that the grouping adds no significant time at all (and takes shorter than the previous checking?)
2. I'm trying to think ahead to what the generic grouping function in c++ would look like. It seems like the distance between the current grouping operator and the generic one is mainly the check ensuring the tensorlists were similar enough. Could we already make this generic by allowing a flag to do this check vs not? Or we could have another util that will just group a list of tensors instead of a nested list.
3. I would prefer if we could already land the generalized version in this PR (and have it live in foreach_utils) if there are no downsides.

**Comment by crcrpar** [comment]
> 1. Is it correct that this PR moved the grouping logic onto CUDA by calling it within the kernel? If so, are the numbers for fused_adam(w) basically showing that the grouping adds no significant time at all (and takes shorter than the previous checking?)

yes. IIUC the difference mainly come from C++ or Python including the setting device guard and handling of map/dict of device and Tensor for grad scaler & found inf.



> * I'm trying to think ahead to what the generic grouping function in c++ would look like. It seems like the distance between the current grouping operator and the generic one is mainly the check ensuring the tensorlists were similar enough. Could we already make this generic by allowing a flag to do this check vs not? Or we could have another util that will just group a list of tensors instead of a nested list.
> * I would prefer if we could already land the generalized version in this PR (and have it live in foreach_utils) if there are no downsides.

I've been played with `torch/csrc/Module.cpp` and am yet to figure out what will suit better with SGD's logic here: https://github.com/pytorch/pytorch/blob/ae3316c16e7cecbee24378ba299f1beab4b60690/torch/optim/sgd.py#L49

I guess having the flag to tell whether or not momentum buffers need initialization will work well, e.g. https://github.com/NVIDIA/apex/blob/7150e20cc3adef34e3f36261a1070fc0882f16a7/apex/optimizers/fused_sgd.py#L121-L136, which I think backward breaking though

File: aten/src/ATen/native/ForeachUtils.h [hunk 5]
```diff
@@ -173,5 +180,140 @@ bool can_use_fast_route(TensorList tensors1, TensorList tensors2, bool does_op_p
...
+              } else {
+                IndicesT indices;
+                indices.reserve(num_tensors);
+                return indices;
+              }
+            }()
+          }
+        }
+      );
+    }
```
Comment (janeyx99) [inline aten/src/ATen/native/ForeachUtils.h:241]: should this logic just be moved before the for loop so it doesn't keep checking with count?

File: aten/src/ATen/native/cuda/FusedAdamWKernel.cu [hunk 6]
```diff
@@ -29,16 +39,40 @@ void _fused_adamw_kernel_cuda_(
...
+  auto device_found_inf_map = init_map(found_inf);
+  const auto nested_tensorlists = [&]() -> std::vector<TensorList> {
+    if (amsgrad) {
+      return {params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs, state_steps};
+    } else {
+      return {params, grads, exp_avgs, exp_avg_sqs, state_steps};
+    }
+  }();
+  OptionalDeviceGuard guard;
+  for (const auto & device_map : group_tensors_by_device_and_scalartype(nested_tensorlists, /* has_state_steps */ true)) {
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/FusedAdamWKernel.cu:52]: Will you be switching this to call the generic foreach util once that's ready? I think that would be ideal (vs having a copy in the fused_adam_utils) and I'd be curious if the perf results would be different.

**Comment by crcrpar** [comment]
@janeyx99 I found this branch a bit too messed up so refactored it into 100007

**Comment by janeyx99** [comment]
Thanks for the update--I'll look there!

**Pull request closed without merging**